crates/swc_plugin_runner/src/load_plugin.rs

```diff
@@ -0,0 +1,152 @@
+                    plugin_path
+                        .file_name()
+                        .and_then(|f| f.to_str())
+                        .expect("Plugin path missing file name"),
+                )
+                .finalize()?;
+
+                // Generate an `ImportObject` from wasi_env
+                let mut import_object = wasi_env.import_object(&module)?;
+
+                // This'll inject few interfaces into plugin's namespace to let plugin
+                // communicate with host
+                let mut env = Exports::new();
+                env.insert("__set_transform_result", set_transform_result_fn_decl);
+                env.insert("__emit_diagnostics", emit_diagnostics_fn_decl);
+
+                env.insert("__mark_fresh_proxy", mark_fresh_fn_decl);
+                env.insert("__mark_parent_proxy", mark_parent_fn_decl);
+                env.insert("__mark_is_builtin_proxy", mark_is_builtin_fn_decl);
+                env.insert("__mark_set_builtin_proxy", mark_set_builtin_fn_decl);
+                env.insert(
+                    "__mark_is_descendant_of_proxy",
+                    mark_is_descendant_of_fn_decl,
+                );
+                env.insert("__mark_least_ancestor", mark_least_ancestor_fn_decl);
+                env.insert(
+                    "__syntax_context_apply_mark_proxy",
+                    syntax_context_apply_mark_fn_decl,
+                );
+                env.insert(
+                    "__syntax_context_remove_mark_proxy",
+                    syntax_context_remove_mark_fn_decl,
+                );
+                env.insert("__syntax_context_outer_proxy", syntax_context_outer_fn_decl);
+
+                import_object.register("env", env);
+                import_object
+            }
+            // Not able to detect wasi version in binary - assume plugin targets
+            // wasm32-unknown-unknown
+            else {
+                imports! {
+                    "env" => {
+                        "__set_transform_result" => set_transform_result_fn_decl,
+                        "__emit_diagnostics" => emit_diagnostics_fn_decl,
+                        "__mark_fresh_proxy" => mark_fresh_fn_decl,
+                        "__mark_parent_proxy" => mark_parent_fn_decl,
+                        "__mark_is_builtin_proxy" => mark_is_builtin_fn_decl,
+                        "__mark_set_builtin_proxy" => mark_set_builtin_fn_decl,
+                        "__mark_is_descendant_of_proxy" => mark_is_descendant_of_fn_decl,
+                        "__mark_least_ancestor" => mark_least_ancestor_fn_decl,
+                        "__syntax_context_apply_mark_proxy" => syntax_context_apply_mark_fn_decl,
+                        "__syntax_context_remove_mark_proxy" => syntax_context_remove_mark_fn_decl,
+                        "__syntax_context_outer_proxy" => syntax_context_outer_fn_decl
+                    }
+                }
+            };
```

Comment:
Rather than if/else you can probably chain the WASI imports along with your custom ones.
Basically, creating your imports:

```rust
                imports! {
                    "env" => {
                        "__set_transform_result" => set_transform_result_fn_decl,
                        "__emit_diagnostics" => emit_diagnostics_fn_decl,
                        "__mark_fresh_proxy" => mark_fresh_fn_decl,
                        "__mark_parent_proxy" => mark_parent_fn_decl,
                        "__mark_is_builtin_proxy" => mark_is_builtin_fn_decl,
                        "__mark_set_builtin_proxy" => mark_set_builtin_fn_decl,
                        "__mark_is_descendant_of_proxy" => mark_is_descendant_of_fn_decl,
                        "__mark_least_ancestor" => mark_least_ancestor_fn_decl,
                        "__syntax_context_apply_mark_proxy" => syntax_context_apply_mark_fn_decl,
                        "__syntax_context_remove_mark_proxy" => syntax_context_remove_mark_fn_decl,
                        "__syntax_context_outer_proxy" => syntax_context_outer_fn_decl
                    }
                }
```

And then detecting if it's a WASI module, and chaining the WASI imports to it with:

https://docs.rs/wasmer/latest/wasmer/trait.ChainableNamedResolver.html#method.chain_front

(that would make the code probably a bit simpler)

Reply:
@syrusakbary Thanks for the tip. Does this https://github.com/swc-project/swc/pull/4012/files#diff-b1fd8abab92d7f7db02dccde2eb021904f6e215ab0fb78553e197ef65fd7a058R102-R118 looks correct? seems it's working as expected at least.

Reply:
Yup, looks correct!